### PR TITLE
Fix #220 on main branch

### LINKED ISSUE #220: Latest version of azure-core (1.21.2) does not work with sanitized recordings
See this commit: https://github.com/microsoft/qdk-python/pull/218/commits/1b9d8d339c306a29cda05c3729a1f95876dee66b

Workaround: https://github.com/microsoft/qdk-python/pull/218/commits/b96eaf472fdbc3cd9eaffd3daf7d32acdad95549

azure-core 1.21.2 checks the http response if the payload size matches the size in the header. since we have a logic in VCR to sanitize the recordings, the size of the payload is not matching

Reported by @vxfield

### PR DESCRIPTION
**Pull request by vxfield**
This PR fixes #220 by [correctly setting the `content-length` HTTP header](https://github.com/microsoft/qdk-python/pull/383/commits/8459cabdd27f2ec0229d6cb15522f1d5d4a74ca2) when sanitizing our test recordings.
It also contains re-recorded unit tests with the correct `content-length` header.

### CODE REVIEW

File: azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_simulated_annealing.yaml [hunk 1]
```diff
@@ -191,113 +189,14 @@ interactions:
...
-        "00000000-0000-0000-0000-000000000000", "status": "Executing", "outputDataFormat":
-        "microsoft.qio-results.v2", "outputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/outputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-SimulatedAnnealing-%252220210101-000000%2522-7857031f-d0c7-11ec-b00f-2816a847b9a3.output.json",
-        "creationTime": "2022-05-11T01:12:49.0442689+00:00", "beginExecutionTime":
-        "2022-05-11T01:12:49.5353146Z", "endExecutionTime": null, "cancellationTime":
-        null, "costEstimate": null, "errorData": null, "isCancelling": false, "tags":
-        [], "access_token": "fake_token"}'
-    headers:
-      content-length:
-      - '1451'
+      - '1037'
```
Comment (anpaz) [inline azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_simulated_annealing.yaml:199]: this is missing a couple of requests, is that expected?

File: azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_simulated_annealing.yaml [hunk 2]
```diff
@@ -312,124 +211,25 @@ interactions:
...
-      - application/json
-      User-Agent:
-      - async-testapp azsdk-python-quantum/0.0.0.1 Python/3.9.12 (Windows-10-10.0.19042-SP0)
-    method: GET
-    uri: https://eastus.quantum.azure.com/v1.0/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myresourcegroup/providers/Microsoft.Quantum/workspaces/myworkspace/jobs/00000000-0000-0000-0000-000000000000
-  response:
-    body:
-      string: '{"containerUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000?sv=PLACEHOLDER&sig=PLACEHOLDER&se=PLACEHOLDER&srt=co&ss=b&sp=racw",
-        "inputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/inputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-SimulatedAnnealing-%252220210101-000000%2522-7857031f-d0c7-11ec-b00f-2816a847b9a3.input.json",
+        "inputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/inputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-SimulatedAnnealing-%252220210101-000000%2522-c08b6f32-3461-11ed-9a0c-6045bdc85d00.input.json",
```
Comment (anpaz) [inline azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_simulated_annealing.yaml:220]: this is also missing.

File: azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_substochastic_monte_carlo.yaml [hunk 3]
```diff
@@ -279,58 +277,25 @@ interactions:
...
-      User-Agent:
-      - async-testapp azsdk-python-quantum/0.0.0.1 Python/3.9.12 (Windows-10-10.0.19042-SP0)
-    method: GET
-    uri: https://eastus.quantum.azure.com/v1.0/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myresourcegroup/providers/Microsoft.Quantum/workspaces/myworkspace/jobs/00000000-0000-0000-0000-000000000000
-  response:
-    body:
-      string: '{"containerUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000?sv=PLACEHOLDER&sig=PLACEHOLDER&se=PLACEHOLDER&srt=co&ss=b&sp=racw",
-        "inputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/inputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-SubstochasticMonteCarlo-%252220210101-000000%2522-816ec5c0-d0c7-11ec-b383-2816a847b9a3.input.json",
+        "inputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/inputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-SubstochasticMonteCarlo-%252220210101-000000%2522-c4336cc5-3461-11ed-9d5f-6045bdc85d00.input.json",
```
Comment (anpaz) [inline azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_substochastic_monte_carlo.yaml:286]: missing parameters.

File: azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_tabu.yaml [hunk 4]
```diff
@@ -245,35 +246,36 @@ interactions:
...
       - '2021-04-10'
     method: GET
-    uri: https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/rawOutputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-Tabu-%252220210101-000000%2522-84f2c2b9-d0c7-11ec-acec-2816a847b9a3.output.json
+    uri: https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/rawOutputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-Tabu-%252220210101-000000%2522-c757123a-3461-11ed-bd9d-6045bdc85d00.output.json
   response:
     body:
-      string: '{"version": "1.0", "configuration": {"1": 1, "0": 1, "2": 0, "3": 1},
-        "cost": -5.0, "solutions": [{"configuration": {"1": 1, "0": 1, "2": 0, "3":
-        1}, "cost": -5.0}], "access_token": "fake_token"}'
+      string: '{"version": "1.0", "configuration": {"0": 1, "1": 1, "2": 0, "3": 1},
```
Comment (anpaz) [inline azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_tabu.yaml:260]: these parameters look different.

File: azure-quantum/tests/unit/aio/recordings/test_streaming_problem_initial_terms.yaml [hunk 5]
```diff
@@ -23,7 +23,7 @@ interactions:
       message: OK
     url: https://login.microsoftonline.com/00000000-0000-0000-0000-000000000000/oauth2/v2.0/token
 - request:
-    body: 'b''{"containerName": "88e6fc91-d0c7-11ec-b9cb-2816a847b9a3"}'''
+    body: 'b''{"containerName": "cae62749-3461-11ed-9e81-6045bdc85d00"}'''
```
Comment (anpaz) [inline azure-quantum/tests/unit/aio/recordings/test_streaming_problem_initial_terms.yaml:26]: wondering if we should also normalize these guids?

File: azure-quantum/tests/unit/recordings/test_job_attachments.yaml [hunk 6]
```diff
@@ -208,33 +208,33 @@ interactions:
...
-        "inputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/inputData",
+        "inputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/inputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=rcw",
         "inputDataFormat": "microsoft.qio.v2", "inputParams": {"params": {}}, "providerId":
         "microsoft", "target": "microsoft.simulatedannealing-parameterfree.cpu", "metadata":
         null, "name": "test_job_attachments", "id": "00000000-0000-0000-0000-000000000000",
         "status": "Waiting", "outputDataFormat": "microsoft.qio-results.v2", "outputDataUri":
-        "https://e2etests.blob.core.windows.net:443/job-00000000-0000-0000-0000-000000000000/outputData?sv=PLACEHOLDER&sig=PLACEHOLDER&se=PLACEHOLDER&srt=co&ss=b&sp=racw",
-        "creationTime": "2022-04-29T17:34:57.254047+00:00", "beginExecutionTime":
+        "https://azuresdkteststo.blob.core.windows.net:443/job-00000000-0000-0000-0000-000000000000/outputData?sv=PLACEHOLDER&sig=PLACEHOLDER&se=PLACEHOLDER&srt=co&ss=b&sp=racw",
```
Comment (anpaz) [inline azure-quantum/tests/unit/recordings/test_job_attachments.yaml:230]: we should probably normalize the storage-name too.

File: azure-quantum/tests/unit/recordings/test_job_submit_honeywell.yaml [hunk 7]
```diff
@@ -215,116 +215,24 @@ interactions:
...
-      x-ms-lease-status:
-      - unlocked
-      x-ms-server-encrypted:
-      - 'true'
-      x-ms-version:
-      - '2020-10-02'
     status:
-      code: 206
-      message: Partial Content
+      code: 403
```
Comment (anpaz) [inline azure-quantum/tests/unit/recordings/test_job_submit_honeywell.yaml:236]: Looks like this job failed. 
The test should have probably failed or at least report a warning. This should run again vs a valid workspace.

File: azure-quantum/tests/unit/recordings/test_job_submit_microsoft_parallel_tempering.yaml [hunk 8]
```diff
@@ -337,25 +342,392 @@ interactions:
...
+        "00000000-0000-0000-0000-000000000000", "status": "Succeeded", "outputDataFormat":
+        "microsoft.qio-results.v2", "outputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/rawOutputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-ParallelTempering-%252220210101-000000%2522-e5f656ed-3460-11ed-91e2-6045bdc85d00.output.json",
+        "creationTime": "2022-09-14T19:10:31.452766+00:00", "beginExecutionTime":
+        "2022-09-14T19:10:28.4686029Z", "endExecutionTime": "2022-09-14T19:10:28.521778Z",
+        "cancellationTime": null, "costEstimate": null, "errorData": null, "isCancelling":
+        false, "tags": [], "access_token": "fake_token"}'
+    headers:
+      content-length:
+      - '1325'
       content-type:
```
Comment (anpaz) [inline azure-quantum/tests/unit/recordings/test_job_submit_microsoft_parallel_tempering.yaml:731]: a bunch of missing requests in this recording.

File: azure-quantum/tests/unit/recordings/test_job_submit_microsoft_population_annealing.yaml [hunk 9]
```diff
@@ -340,25 +341,394 @@ interactions:
...
+        "id": "00000000-0000-0000-0000-000000000000", "status": "Succeeded", "outputDataFormat":
+        "microsoft.qio-results.v2", "outputDataUri": "https://mystorage.blob.core.windows.net/job-00000000-0000-0000-0000-000000000000/rawOutputData?sv=PLACEHOLDER&sr=b&sig=PLACEHOLDER&se=PLACEHOLDER&sp=r&rscd=attachment%3B%20filename%3DTest-PopulationAnnealing-%252220210101-000000%2522-ebad6cfa-3460-11ed-bc1d-6045bdc85d00.output.json",
+        "creationTime": "2022-09-14T19:10:40.7809122+00:00", "beginExecutionTime":
+        "2022-09-14T19:10:40.0750431Z", "endExecutionTime": "2022-09-14T19:10:40.230891Z",
+        "cancellationTime": null, "costEstimate": null, "errorData": null, "isCancelling":
+        false, "tags": [], "access_token": "fake_token"}'
+    headers:
+      content-length:
+      - '1334'
       content-type:
```
Comment (anpaz) [inline azure-quantum/tests/unit/recordings/test_job_submit_microsoft_population_annealing.yaml:732]: missing recordings.

File: azure-quantum/tests/unit/recordings/test_workspace_job_quotas.yaml [hunk 10]
```diff
@@ -48,34 +48,23 @@ interactions:
...
   response:
     body:
       string: '{"value": [{"dimension": "combined_job_hours", "scope": "Workspace",
-        "providerId": "Microsoft", "utilization": 0.06583260363888885, "holds": 0.0,
-        "limit": 20.0, "period": "Monthly"}, {"dimension": "combined_job_hours", "scope":
-        "Subscription", "providerId": "Microsoft", "utilization": 0.5987412553611111,
-        "holds": 0.01666666666666672, "limit": 1000.0, "period": "Monthly"}, {"dimension":
-        "concurrent_cpu_jobs", "scope": "Workspace", "providerId": "Microsoft", "utilization":
-        0.0, "holds": 0.0, "limit": 5.0, "period": "None"}, {"dimension": "concurrent_fpga_jobs",
+        "providerId": "Microsoft", "utilization": 0.009026114250000002, "holds": 0.0,
```
Comment (anpaz) [inline azure-quantum/tests/unit/recordings/test_workspace_job_quotas.yaml:57]: Assuming the missing dimensions is on purpose?

**Review by ricardo-espinoza — APPROVED** [review]

File: azure-quantum/tests/unit/recordings/test_quil_syntax_error.yaml [hunk 11]
```diff
@@ -1,38 +1,38 @@
 interactions:
 - request:
-    body: client_id=PLACEHOLDER&grant_type=client_credentials&client_info=1&client_secret=PLACEHOLDER&claims=PLACEHOLDER&scope=https%3A%2F%2Fquantum.microsoft.com%2F.default
+    body: client_id=PLACEHOLDER&grant_type=client_credentials&client_info=1&client_secret=+EDw8Q~cbkbeadF.mrC6bAXxFEspQDblyNyuRkcfT&claims=PLACEHOLDER&scope=https%3A%2F%2Fquantum.microsoft.com%2F.default
```
Comment (anpaz) [inline azure-quantum/tests/unit/recordings/test_quil_syntax_error.yaml:3]: leaked secret.

File: azure-quantum/tests/unit/recordings/test_streaming_problem_initial_terms.yaml [hunk 12]
```diff
@@ -1,6 +1,6 @@
 interactions:
 - request:
-    body: client_id=PLACEHOLDER&grant_type=client_credentials&client_info=1&client_secret=PLACEHOLDER&claims=PLACEHOLDER&scope=https%3A%2F%2Fquantum.microsoft.com%2F.default
+    body: client_id=PLACEHOLDER&grant_type=client_credentials&client_info=1&client_secret=+EDw8Q~cbkbeadF.mrC6bAXxFEspQDblyNyuRkcfT&claims=PLACEHOLDER&scope=https%3A%2F%2Fquantum.microsoft.com%2F.default
```
Comment (anpaz) [inline azure-quantum/tests/unit/recordings/test_streaming_problem_initial_terms.yaml:3]: leaked secret

**Review by anpaz — COMMENTED** [review]
Still need to review more, but left a couple of observations.

File: azure-quantum/tests/unit/recordings/test_job_submit_honeywell.yaml (hunk 7 above)
Comment (vxfield) [reply azure-quantum/tests/unit/recordings/test_job_submit_honeywell.yaml:236]: Oh, this should have been caught by the unit tests asserts, but unfortunately there was [some logic to suppress the error of not having the provider in the workspace and make it just a warning](https://github.com/microsoft/qdk-python/blob/179fb04f7b2ec5b87b26551722400f0dd437a575/azure-quantum/tests/unit/test_target.py#L235).
I will remove the logic to let the test fail in those cases.

File: azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_substochastic_monte_carlo.yaml (hunk 3 above)
Comment (vxfield) [reply azure-quantum/tests/unit/aio/recordings/test_job_submit_microsoft_substochastic_monte_carlo.yaml:286]: I'm assuming the test case, the Python SDK or Azure Quantum backend may have changed to cause this difference.
I'm also counting on the unit tests asserts to make sure this is no issue.

**Comment by vxfield** [comment]
Addressed all comments on #385

**Pull request merged**